src/mlpack/methods/ann/layer/weight_norm.hpp

```diff
@@ -0,0 +1,204 @@
+#include "layer_types.hpp"
+
+#include "../visitor/delete_visitor.hpp"
+#include "../visitor/delta_visitor.hpp"
+#include "../visitor/output_parameter_visitor.hpp"
+#include "../visitor/reset_visitor.hpp"
+#include "../visitor/weight_size_visitor.hpp"
+#include "../visitor/weight_set_visitor.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Declaration of the WeightNorm layer class. The layer reparameterizes
+ * the weight vectors in a neural network, decoupling the length of those weight
+ * vectors from their direction. This reparameterization does not introduce any
+ * dependencies between the examples in a mini-batch.
+ *
+ * This class will be a wrapper around existing layers. It will just modify the
+ * calculation and updation of weights of the layer.
+ *
+ * For more information, refer to the following paper,
+ *
+ * @code
+ * @inproceedings{Salimans2016WeightNorm,
+ *   title = {Weight Normalization: A Simple Reparameterization to Accelerate
+ *            Training of Deep Neural Networks},
+ *   author = {Tim Salimans, Diederik P. Kingma},
+ *   booktitle = {Neural Information Processing Systems 2016},
+ *   year = {2016}
+ * }
+ * @endcode
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam CustomLayers Additional custom layers that can be added.
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat,
+  typename... CustomLayers
+>
+class WeightNorm
+{
+ public:
+  /**
+   * Create the WeightNorm layer object.
+   *
+   * @param layer The layer whose weights are needed to be normalized.
+   */
+  WeightNorm(LayerTypes<CustomLayers...> layer = LayerTypes<CustomLayers...>());
+
+  //! Destructor to release allocated memory.
+  ~WeightNorm();
+
+  /**
+   * Reset the layer parameters.
+   */
+  void Reset();
+
+  /**
+   * Forward pass of the WeightNorm layer. Calculates the weights of the
+   * wrapped layer from the parameter vector v and the scalar parameter g.
+   * It then calulates the output of the wrapped layer from the calculated
+   * weights.
+   *
+   * @param input Input data for the layer.
+   * @param output Resulting output activations.
+   */
+  template<typename eT>
+  void Forward(arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Backward pass through the layer. This function calls the Backward()
+   * function of the wrapped layer.
+   *
+   * @param input The input activations.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& input,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activations and
+   * weights of the wrapped layer.
```

Comment:
This should be like `... using the output delta, input activations and the weights ... `